src/Fakta/Agent.fs

```diff
@@ -1,78 +1,389 @@
+/// Members returns the known gossip members. The WAN flag can be used to query a server for WAN members.
+let members (state : FaktaState) (wan : bool) : Async<Choice<AgentMember list, Error>> =
+  let getResponse = Impl.getResponse state "Fakta.Agent.members"
+  let req =
+    UriBuilder.ofAgent state.config "members" 
+    |> UriBuilder.uri
+    |> basicRequest Get
+    |> withConfigOpts state.config
+  async {
+  let! resp, dur = Duration.timeAsync (fun () -> getResponse req)
+  match resp with
+  | Choice1Of2 resp ->
+    use resp = resp
+    if not (resp.StatusCode = 200 || resp.StatusCode = 404) then
+      return Choice2Of2 (Message (sprintf "unknown response code %d" resp.StatusCode))
+    else
+      match resp.StatusCode with
+      | 404 -> return Choice2Of2 (Message "agent.members")
+      | _ ->
+        let! body = Response.readBodyAsString resp
+        let  item = if body = "" then [] else Json.deserialize (Json.parse body)
+        return Choice1Of2 (item)
 
-/// PassTTL is used to set a TTL check to the passing state
-let passTTL (state : FaktaState) (checkId : string) (note : string) : Async<Choice<unit, Error>> =
-  raise (TBD "TODO")
+  | Choice2Of2 exx ->
+    return Choice2Of2 (Error.ConnectionFailed exx)
+}
 
 /// Self is used to query the agent we are speaking to for information about itself
 let self (state : FaktaState) : Async<Choice<Map<string, Map<string, Chiron.Json>>, Error>> =
-  raise (TBD "TODO")
+  let getResponse = Impl.getResponse state "Fakta.Agent.self"
+  let req =
+    UriBuilder.ofAgent state.config "self" 
+    |> UriBuilder.uri
+    |> basicRequest Get
+    |> withConfigOpts state.config
+  async {
+  let! resp, dur = Duration.timeAsync (fun () -> getResponse req)
+  match resp with
+  | Choice1Of2 resp ->
+    use resp = resp
+    if not (resp.StatusCode = 200 || resp.StatusCode = 404) then
+      return Choice2Of2 (Message (sprintf "unknown response code %d" resp.StatusCode))
+    else
+      match resp.StatusCode with
+      | 404 -> return Choice2Of2 (Message "agent.self")
+      | _ ->
+        let! body = Response.readBodyAsString resp
+        let  item = if body = "" then Map.empty else Json.deserialize (Json.parse body)
+        return Choice1Of2 (item)
+
+  | Choice2Of2 exx ->
+    return Choice2Of2 (Error.ConnectionFailed exx)
+}
+
+/// NodeName is used to get the node name of the agent
+let nodeName (state : FaktaState) : Async<Choice<string, Error>> =
+  async {    
+    match Async.RunSynchronously (self state) with
+    | Choice1Of2 map -> 
+        let r = match map.ContainsKey("Config") with
+                | true ->
+                    let config = map.Item "Config"
+                    match config.ContainsKey("NodeName") with
+                    | true -> 
+                        let nodeNameJson = config.Item "NodeName"
+                        let nodeName = Json.format nodeNameJson
+                        nodeName
+                    | false -> "Node name not found"
```

Comment:
This is wrong; if the node name wasn't found it should be specified as a value in the `Error` discriminated union. This code means; "The node name is 'Node name not found'", which is false.